### PR TITLE
Add codespell pre-commit hook

### LINKED ISSUE #378: Add pre-commit hooks for spell checking
## Problem

We don't have Git hooks to check for spelling mistakes in code or in commit messages.

## Description

`pre-commit` has several hooks that could provide this check. Visit https://pre-commit.com/hooks.html and filter by the words 'spell' or 'typo' to see a list of options. Finding a popular, maintained one in the list and using it would improve the spelling situation considerably.

## Alternatives

Relying on spell check by our IDEs and text editors is okay but it doesn't always work. Just yesterday, I made a commit called "Update pacakges" even though my IDE showed a wavy underline under the word.

## References

Here are some of the most starred options from pre-commit's list:
- https://github.com/codespell-project/codespell (1.1k)
- https://github.com/crate-ci/typos (985)
- https://github.com/streetsidesoftware/cspell-cli (25, but uses https://github.com/streetsidesoftware/cspell (710))
- https://github.com/hcodes/yaspeller (620)

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Fixes

Fixes #378 by @dhruvkb 

## Description

This PR does the following:

1. Consolidates Python-based tool configuration into a single pyproject.toml. This file is auto-detected by all the relevant tools. Using it reduces the number of disparate files with varying naming conventions (dictated by the tools) at the repository root.
2. Adds and configures codespell, including additional documentation to assist with ongoing maintenance of the configuration.

Codespell seems fine enough and works pretty well. ~However, I would like to also explore using cspell. It has far greater flexibility in terms of the supported dictionaries and its configuration is significantly less obscure than Codespells. I'm going to leave this PR as a draft until I've tried cspell. If I decide to go with cspell, I'll open a new PR and keep the configuration consolidation I did for other Python-based tools in this PR. That seems useful regardless of which spell checker we choose.~ I tried out cspell and while it probably works great if you have a small project with little prose, it complains about every unknown word. At a glance I could not find anything in cspell's documentation about how to ignore unknown words, I suspect because its strategy for matching words is very different. Codespell is going to be much easier to maintain in the long run. It isn't perfect and will probably miss some things that cspell would catch, but it isn't worth the tremendous task of configuration and maintenance to use cspell. Keep in mind that "unknown words" includes most Vue related terms, all maintainer GitHub usernames (referenced frequently in project documentation), and so on and so on. In any case, Codespell is a good start. If we want to switch tools in the future we can do so. For now this gets us at least catching basic misspellings.

## Testing Instructions

Checkout the branch and run `just lint codespell`. Confirm there are no errors. Add a misspelling to a relevant file (one that isn't ignored) and confirm that codespell catches it when running `just lint codespell`. Also, review the list of ignored files and confirm that it's comprehensive and does not accidentally include files that _should_ be checked.

Look through the spelling changes and confirm that there are no problems with the fixes I've applied. There is a mix of automatically applied fixes that codespell generated as well as fixes I had to do manually (for when there were multiple suggestions for a misspelled word).

Finally, read through the configuration documentation and confirm that it makes sense, is clear, and can be used effectively to maintain the configuration moving forward.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [N/A] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: catalog/dags/providers/provider_api_scripts/metropolitan_museum.py
Comment (dhruvkb) [inline catalog/dags/providers/provider_api_scripts/metropolitan_museum.py:71]: This is unfortunate 😄.

File: catalog/tests/dags/providers/provider_api_scripts/test_wikimedia_commons.py
Comment (dhruvkb) [inline catalog/tests/dags/providers/provider_api_scripts/test_wikimedia_commons.py:419]: Considering that this is the name of an artwork, a `# noqa` (or equivalent) would be more appropriate here.

File: documentation/meta/ci_cd/proof_of_functionality.md
Comment (dhruvkb) [inline documentation/meta/ci_cd/proof_of_functionality.md:26]: I'm a bit undecided on whether we should limit the dictionary to US English. On one hand, consistency 👍. On the other, these spellings look weird to me 😄.

File: documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20230309-implementation_plan_sensitive_terms_list.md
Comment (dhruvkb) [inline documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20230309-implementation_plan_sensitive_terms_list.md:95]: TIL!

File: frontend/test/unit/specs/utils/decode-data.spec.js
Comment (dhruvkb) [inline frontend/test/unit/specs/utils/decode-data.spec.js:40]: Maybe an ESLint rule for test names to start with "should"/"should not"/"does"/"does not"? I don't know if that seems excessively restrictive without much additional benefit.

File: catalog/tests/dags/common/test_resources/image_table_sample.tsv
Comment (dhruvkb) [inline catalog/tests/dags/common/test_resources/image_table_sample.tsv:18]: I'd argue that this file should be excluded from spell-check because we do not control the accuracy of spelling of media titles, tags and descriptions ingested from sources.

Comment (dhruvkb) [reply catalog/tests/dags/common/test_resources/image_table_sample.tsv:18]: Also this change seems to be in the name of a media creator, which is wrong.

File: .codespell/README.md
Comment (dhruvkb) [inline .codespell/README.md:1]: IMO, this contains a lot of valuable information that might be better moved to a docs page (linked to from this `README.md` file) than here where it is not searchable.

File: .codespell/dictionary.txt
Comment (dhruvkb) [inline .codespell/dictionary.txt:1]: Nice, I was also thinking we could add these (if not already) and document this file in the Sphinx docs so that we, and new contributors, know where to add more definitions in the future.

```
wordpress->WordPress
typescript->TypeScript
javascript->JavaScript
django->Django
```

File: .pre-commit-config.yaml
Comment (dhruvkb) [inline .pre-commit-config.yaml:130]: There is a spelling correction in a test resource file in this PR, that might not be necessary with this skip.

Side-note/tangent: I'm not able to mentally track which file/directory is skipped by which tool so it might be a good idea to either have some docs reference for skipping or have the skips be defined in a common place.

**Review by dhruvkb — CHANGES_REQUESTED** [review]
LGTM, but I have a number of suggestions so I think blocking merge to have these discussions, is justified. Also, I'm quite shocked at how many spelling mistakes we've committed!

File: catalog/tests/dags/providers/provider_api_scripts/test_wikimedia_commons.py
Comment (sarayourfriend) [reply catalog/tests/dags/providers/provider_api_scripts/test_wikimedia_commons.py:419]: Oops! This was one that got automatically caught. codespell doesn't have `noqa`, but we can add it to the `ignore_lines.txt` :+1:

File: frontend/test/unit/specs/utils/decode-data.spec.js
Comment (sarayourfriend) [reply frontend/test/unit/specs/utils/decode-data.spec.js:40]: I like it from a perspective of encouraging descriptive (and "self documenting") test names, but it would require a _lot_ of changes in our tests, I think... we've got a lot of tests that aren't formatted that way.

File: catalog/tests/dags/common/test_resources/image_table_sample.tsv
Comment (sarayourfriend) [reply catalog/tests/dags/common/test_resources/image_table_sample.tsv:18]: Yes! I missed this one from the ignore files list :+1:

File: .codespell/dictionary.txt
Comment (sarayourfriend) [reply .codespell/dictionary.txt:1]: Unfortunately codespell does not match case on the left-hand side :cry: Hence the issue with the `afterAll` Jest hook getting confused with `afterall->after all`. This Openverse one doesn't even work anyway, or else there would be a lot of places to add ignores for.

File: .pre-commit-config.yaml
Comment (sarayourfriend) [reply .pre-commit-config.yaml:130]: I got confused about this today as well with the ESLint configuration. Agreed that additional documentation here would be useful.

File: documentation/meta/ci_cd/proof_of_functionality.md
Comment (sarayourfriend) [reply documentation/meta/ci_cd/proof_of_functionality.md:26]: The default dictionary, IIRC, doesn't enforce US spellings in most cases. Rigourous isn't an accepted spelling anywhere, as far as I can see. Cambridge Dictionary does not mention alternative spellings for UK, at least: https://dictionary.cambridge.org/dictionary/english/rigour. Wikitionary says it's a misspelling as well, or at best an archaic version: https://en.wiktionary.org/wiki/rigourous#English

Anyway, I also agree we don't need to worry about enforcing US spellings. I would personally find it annoying to need to switch between it depending on the contexts (work vs everywhere else in my life including school), particularly when for _this_ context it matters very little if at all and in other contexts is more important.

All of this is TIL, FWIW. The French made English so confusing :rofl:

**Comment by sarayourfriend** [comment]
Thank you for the careful review, Dhruv. I'll make updates to this tomorrow!

File: documentation/meta/ci_cd/proof_of_functionality.md
Comment (dhruvkb) [reply documentation/meta/ci_cd/proof_of_functionality.md:26]: That's my bad, I didn't look it up, just added "ous" to "rigour".

Comment (sarayourfriend) [reply documentation/meta/ci_cd/proof_of_functionality.md:26]: I had to look it up too :) It's so often English defies all my expectations lol!

**Review by obulat — APPROVED** [review]
Really nice clean up, both in terms of combining the configuration into a single `toml` file, and fixing all of the spelling errors. Thank you for detailed documentation, too!

**Review by dhruvkb — APPROVED** [review]
Seems like all my suggestions have been addressed, thank you! LGTM, LGTM!

**Pull request merged**
